File: unit-testing/java-good-coverage/src/test/java/com/calculator/steps/CalculatorSteps.java
package com.calculator.steps;

import com.calculator.Calculator;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.When;
import io.cucumber.java.en.Then;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Step definitions for Calculator Cucumber tests
 */
public class CalculatorSteps {

    private Calculator calculator;
    private double result;
    private Exception exception;

    @Given("I have a calculator")
    public void i_have_a_calculator() {
        calculator = new Calculator();
        exception = null;
    }

    @When("I add {int} and {int}")
    public void i_add_and(Integer num1, Integer num2) {
        result = calculator.add(num1, num2);
    }

    @When("I subtract {int} from {int}")
    public void i_subtract_from(Integer num2, Integer num1) {
        result = calculator.subtract(num1, num2);
    }

    @When("I multiply {int} and {int}")
    public void i_multiply_and(Integer num1, Integer num2) {
        result = calculator.multiply(num1, num2);
    }

    @When("I divide {int} by {int}")
    public void i_divide_by(Integer num1, Integer num2) {
        try {
            result = calculator.divide(num1, num2);
        } catch (ArithmeticException e) {
            exception = e;
        }
    }

    @When("I raise {int} to the power of {int}")
    public void i_raise_to_the_power_of(Integer base, Integer exponent) {
        result = calculator.power(base, exponent);
    }

    @When("I calculate the square root of {double}")
    public void i_calculate_the_square_root_of(Double number) {
        try {
            result = calculator.squareRoot(number);
        } catch (IllegalArgumentException e) {
            exception = e;
        }
    }

    @Then("the result should be {double}")
    public void the_result_should_be(Double expected) {
        assertEquals(expected, result, 0.001);
    }

    @Then("an ArithmeticException should be thrown")
    public void an_arithmetic_exception_should_be_thrown() {
        assertNotNull(exception, "Expected ArithmeticException to be thrown");
        assertTrue(exception instanceof ArithmeticException);
    }

    @Then("an IllegalArgumentException should be thrown")
    public void an_illegal_argument_exception_should_be_thrown() {
        assertNotNull(exception, "Expected IllegalArgumentException to be thrown");
        assertTrue(exception instanceof IllegalArgumentException);
    }
}

File: unit-testing/python-good-coverage/features/steps/calculator_steps.py
"""Step definitions for Calculator Behave tests."""
from behave import given, when, then
from calculator.calculator import Calculator


@given('I have a calculator')
def step_impl(context):
    """Initialize calculator and exception tracking."""
    context.calculator = Calculator()
    context.exception = None
    context.result = None


@when('I add {num1:d} and {num2:d}')
def step_impl(context, num1, num2):
    """Add two numbers."""
    context.result = context.calculator.add(num1, num2)


@when('I subtract {num2:d} from {num1:d}')
def step_impl(context, num2, num1):
    """Subtract second number from first."""
    context.result = context.calculator.subtract(num1, num2)


@when('I multiply {num1:d} and {num2:d}')
def step_impl(context, num1, num2):
    """Multiply two numbers."""
    context.result = context.calculator.multiply(num1, num2)


@when('I divide {num1:d} by {num2:d}')
def step_impl(context, num1, num2):
    """Divide first number by second."""
    try:
        context.result = context.calculator.divide(num1, num2)
    except ArithmeticError as e:
        context.exception = e


@when('I raise {base:d} to the power of {exponent:d}')
def step_impl(context, base, exponent):
    """Calculate power of a number."""
    context.result = context.calculator.power(base, exponent)


@when('I calculate the square root of {number:f}')
def step_impl(context, number):
    """Calculate square root of a number."""
    try:
        context.result = context.calculator.square_root(number)
    except ValueError as e:
        context.exception = e


@when('I calculate the square root of {number:d}')
def step_impl(context, number):
    """Calculate square root of a number (integer input)."""
    try:
        context.result = context.calculator.square_root(number)
    except ValueError as e:
        context.exception = e


@then('the result should be {expected:f}')
def step_impl(context, expected):
    """Verify the result matches expected value."""
    assert context.result is not None, "Result was not set"
    assert abs(context.result - expected) < 0.001, \
        f"Expected {expected}, but got {context.result}"


@then('an ArithmeticError should be thrown')
def step_impl(context):
    """Verify ArithmeticError was thrown."""
    assert context.exception is not None, "Expected ArithmeticError to be thrown"
    assert isinstance(context.exception, ArithmeticError), \
        f"Expected ArithmeticError, but got {type(context.exception)}"


@then('a ValueError should be thrown')
def step_impl(context):
    """Verify ValueError was thrown."""
    assert context.exception is not None, "Expected ValueError to be thrown"
    assert isinstance(context.exception, ValueError), \
        f"Expected ValueError, but got {type(context.exception)}"
